In [1]:
import pandas as pd
import numpy as np
from category_encoders import BinaryEncoder

# 1. Load data

In [157]:
# set data paths
can_dataset_path_for_D_free = 'dataset/Pre_train_D_attack_free.csv'
can_dataset_path_for_S_free = 'dataset/Pre_train_S_attack_free.csv'
can_dataset_path_for_D_1 = 'dataset/Pre_train_D_attack_1.csv'
can_dataset_path_for_D_2 = 'dataset/Pre_train_D_attack_2.csv'


# data load (Driving state, Stopping state)
data_D_free= pd.read_csv(
    can_dataset_path_for_D_free,
    engine = 'python',
    encoding = 'cp949'
)

data_S_free= pd.read_csv(
    can_dataset_path_for_S_free,
    engine = 'python',
    encoding = 'cp949'
)

data_D_attack_1= pd.read_csv(
    can_dataset_path_for_D_1,
    engine = 'python',
    encoding = 'cp949'
)

data_D_attack_2= pd.read_csv(
    can_dataset_path_for_D_2,
    engine = 'python',
    encoding = 'cp949'
)

# 2. attack free data EDA

In [3]:
data_D_free.head()

Timestamp Arbitration_ID  DLC                     Data   Class
0  1.597708e+09            260    8  06 25 05 30 FF CF 71 55  Normal
1  1.597708e+09            329    8  4A C5 7E 8C 31 2D 01 10  Normal
2  1.597708e+09            38D    8  00 00 49 00 90 7F FE 01  Normal
3  1.597708e+09            420    8  50 1E 00 C8 FC 4F 6A 00  Normal
4  1.597708e+09            421    8  FE 07 00 FF E3 7F 00 52  Normal

In [4]:
data_S_free.head()

Timestamp Arbitration_ID  DLC                     Data   Class
0  1.597708e+09            391    8  00 00 00 00 00 00 04 77  Normal
1  1.597708e+09            251    8  F0 03 A3 EE 00 F3 07 80  Normal
2  1.597708e+09            2B0    6        1D 00 00 07 83 AE  Normal
3  1.597708e+09            4F1    4              20 00 60 41  Normal
4  1.597708e+09            50C    8  40 0E 00 08 06 10 00 D2  Normal

In [5]:
print('## Shape of Driving state ##')
print('number of data : ', data_D_free.shape[0])
print('number of feature : ', data_D_free.shape[1],'\n')
print('## Data types of Driving state ##')
print(data_D_free.dtypes, '\n')
print('## the number of total IDs in Driving state ##')
print('number of IDs : ',len(data_D_free['Arbitration_ID'].unique()),'\n')

## Shape of Driving state ##
number of data :  179346
number of feature :  5 

## Data types of Driving state ##
Timestamp         float64
Arbitration_ID     object
DLC                 int64
Data               object
Class              object
dtype: object 

## the number of total IDs in Driving state ##
number of IDs :  73 



In [6]:
print('## Shape of Stopping state ##')
print('number of data : ', data_S_free.shape[0])
print('number of feature : ', data_S_free.shape[1],'\n')
print('## Data types of Stopping state ##')
print(data_S_free.dtypes, '\n')
print('## the number of total IDs in Stopping state ##')
print('number of IDs : ',len(data_S_free['Arbitration_ID'].unique()),'\n')

## Shape of Stopping state ##
number of data :  180686
number of feature :  5 

## Data types of Stopping state ##
Timestamp         float64
Arbitration_ID     object
DLC                 int64
Data               object
Class              object
dtype: object 

## the number of total IDs in Stopping state ##
number of IDs :  73 



In [14]:
print('## value count of IDs in Driving state ##')
data_D_free['Arbitration_ID'].value_counts()

## value count of IDs in Driving state ##


260    7466
153    7466
220    7466
251    7465
367    7465
       ... 
5B0      75
043      75
5BE      74
07F      74
57F      37
Name: Arbitration_ID, Length: 73, dtype: int64

In [8]:
print('## value count of IDs ##')
data_S_free['Arbitration_ID'].value_counts()

## value count of IDs ##


2B0    7531
251    7531
164    7531
356    7531
366    7531
       ... 
5B0      76
043      75
5BE      75
07F      75
57F      37
Name: Arbitration_ID, Length: 73, dtype: int64

In [29]:
IDs= data_D_free['Arbitration_ID'].value_counts()

In [86]:
print('74-- IDs : ',IDs[IDs>7000].index.sort_values().values)
print('37-- IDs : ',IDs[(IDs >= 3000)&(IDs <= 7000)].index.sort_values().values)
print('14-- IDs : ',IDs[(IDs >= 1300)&(IDs <= 3000)].index.sort_values().values)
print('10-- IDs : ',IDs[(IDs >= 1000)&(IDs <= 1400)].index.sort_values().values)
print('7-- IDs : ',IDs[(IDs >= 700)&(IDs <= 1000)].index.sort_values().values)
print('6-- IDs : ',IDs[(IDs >= 400)&(IDs <= 700)].index.sort_values().values)
print('3-- IDs : ',IDs[(IDs >= 200)&(IDs <= 400)].index.sort_values().values)
print('Else IDs(150>) : ',IDs[IDs <= 200].index.sort_values().values)

74-- IDs :  ['130' '140' '153' '164' '220' '251' '260' '2B0' '329' '340' '356' '366'
 '367' '368']
37-- IDs :  ['381' '386' '387' '389' '38D' '391' '394' '420' '421' '453' '470' '47F'
 '4F1']
14-- IDs :  ['436' '485' '48A' '490' '492' '58B']
10-- IDs :  ['484']
7-- IDs :  ['42D' '479' '495' '500' '507' '50C' '520' '53E' '541' '568']
6-- IDs :  ['52A']
3-- IDs :  ['410' '412' '44E' '483' '48C' '49F' '4A9' '4C9' '4CB' '50A' '50E' '53B'
 '53F' '544' '553' '559' '572' '593' '5A6' '5CD']
Else IDs(150>) :  ['043' '07F' '4A2' '4A7' '563' '57F' '5B0' '5BE']


In [88]:
IDs_= data_S_free['Arbitration_ID'].value_counts()

In [143]:
print('75-- IDs : ',IDs_[IDs_>7000].index.sort_values().values, 'time interval : 25')
print('37-- IDs : ',IDs_[(IDs_ >= 3000)&(IDs_ <= 7000)].index.sort_values().values, 'time interval : 50')
print('15-- IDs : ',IDs_[(IDs_ >= 1300)&(IDs_ <= 3000)].index.sort_values().values, 'time interval : 130')
print('10-- IDs : ',IDs_[(IDs_ >= 1000)&(IDs_ <= 1400)].index.sort_values().values, 'time interval : 170')
print('7-- IDs : ',IDs_[(IDs_ >= 700)&(IDs_ <= 1000)].index.sort_values().values, 'time interval : 240')
print('6-- IDs : ',IDs_[(IDs_ >= 400)&(IDs_ <= 700)].index.sort_values().values)
print('3-- IDs : ',IDs_[(IDs_ >= 200)&(IDs_ <= 400)].index.sort_values().values, 'time interval : 500')
print('Else IDs(150>) : ',IDs_[IDs_ <= 200].index.sort_values().values)

75-- IDs :  ['130' '140' '153' '164' '220' '251' '260' '2B0' '329' '340' '356' '366'
 '367' '368'] time interval : 25
37-- IDs :  ['381' '386' '387' '389' '38D' '391' '394' '420' '421' '453' '470' '47F'
 '4F1'] time interval : 50
15-- IDs :  ['436' '485' '48A' '490' '492' '58B'] time interval : 75
10-- IDs :  ['484']
7-- IDs :  ['42D' '479' '495' '500' '507' '50C' '520' '53E' '541' '568'] time interval : 100
6-- IDs :  []
3-- IDs :  ['410' '412' '44E' '483' '48C' '49F' '4A9' '4C9' '4CB' '50A' '50E' '52A'
 '53B' '53F' '544' '553' '559' '572' '593' '5A6' '5CD']
Else IDs(150>) :  ['043' '07F' '4A2' '4A7' '563' '57F' '5B0' '5BE']


Most of IDs in driving state and stopping state has same elements except '52A'
I select main id to 130, 381, 436, 42D, 410

In [148]:
data_D_free[data_D_free['Arbitration_ID'].isin(['410'])]['Data']

193       60 12 40 00 01 00 00 00
676       60 12 40 00 01 00 00 00
1154      60 12 40 00 01 00 00 00
1632      60 12 40 00 01 00 00 00
2113      60 12 40 00 01 00 00 00
                   ...           
177007    60 12 40 00 01 00 00 00
177487    60 12 40 00 01 00 00 00
177971    60 12 40 00 01 00 00 00
178450    60 12 40 00 01 00 00 00
178930    60 12 40 00 01 00 00 00
Name: Data, Length: 373, dtype: object

Data : data[0], data[1] , ... , data[6], data[7]
- feature of ID 130
--  data[1] is 80 or 7F
--  data[3] is 80 or 7F or 81
--  data[4], data[5] is 00
--  data[6] have order(00 to 0F)

In [105]:
data_D_free[data_D_free['Arbitration_ID'].isin(['381'])]['Data']

19        80 28 42 00 00 EE 12 05
65        80 F0 41 00 00 EF 0E 05
114       80 A8 41 00 00 F0 57 05
169       80 70 41 00 00 F1 F5 05
216       80 48 41 00 00 F2 17 05
                   ...           
179141    80 20 41 00 00 7E DB 05
179185    80 20 41 00 00 7F 97 05
179234    80 20 41 00 00 80 D6 05
179276    80 48 41 00 00 81 E8 05
179329    80 30 41 00 00 82 5A 05
Name: Data, Length: 3733, dtype: object

- feature of ID 381
--  data[0] is 80
--  data[1], data[2] is very different (maybe this sensor is driving system)
--  data[3], data[4] is 00
--  data[5] have order(00 to FF)
--  data[7] is 05

In [114]:
# data_D_free[data_D_free['Arbitration_ID'].isin(['436'])]['Data']
# data_D_free[data_D_free['Arbitration_ID'].isin(['42D'])]['Data']
# data_D_free[data_D_free['Arbitration_ID'].isin(['410'])]['Data']

- feature of ID 436 in D and S
--  all data is 00 00 00 04
- feature of ID 42D in D and S
--  all data is 00 00 00 00 00 D0 07 FA
- feature of ID 410 in D and S
--  all data is 60 12 40 00 01 00 00 00

75-- IDs :
-- perfect order : ['130' '140' '153' '164' '220' '251' '260' '340' ]
-- atypical order : [ '329' '366' '368' ]
-- no order : ['2B0' '356' '367' ]
37-- IDs :
-- perfect order : ['381' '387' '391' '420' '47F' '4F1']
-- atypical order : ['386' '38D' '394' '470' ]
-- no order : ['389' '453' ]

result : some IDs has perfect order such as 00, 01, ... , FE, FF. some of them are same code like ID 436
it is good chance to cheat IDS



In [154]:
print('75-- IDs ', 'time interval : 25')
print('37-- IDs ','time interval : 50')
print('15-- IDs ', 'time interval : 130')
print('10-- IDs ','time interval : 170')
print('7-- IDs ',' time interval : 240')
print('3-- IDs ',' time interval : 500')

75-- IDs  time interval : 25
37-- IDs  time interval : 50
15-- IDs  time interval : 130
10-- IDs  time interval : 170
7-- IDs   time interval : 240
3-- IDs   time interval : 500


# 3. attack data EDA

Attack: Attack traffic injected. Four types of attacks are included -- Flooding, Spoofing, Replay, Fuzzing.

- Flooding: Flooding attack aims to consume CAN bus bandwidth by sending a massive number of messages. ( ID and data are 000 )

- Spoofing: CAN messages are injected to control certain desired function. ( normal ID + fixed data )

- Replay: Replay attack is to extract normal traffic at a specific time and replay (inject) it into the CAN bus. (ctrl+C , ctrl+v)

- Fuzzing: Random messages are injected to cause unexpected behavior of the vehicle. (normal ID +random 7 ID and random data)

In [158]:
data_D_attack_1.head()

Timestamp Arbitration_ID  DLC                     Data   Class SubClass
0  1.597760e+09            153    8  20 A1 10 FF 00 FF 50 1F  Normal   Normal
1  1.597760e+09            220    8  13 24 7F 60 05 FF BF 10  Normal   Normal
2  1.597760e+09            507    4              08 00 00 01  Normal   Normal
3  1.597760e+09            356    8  00 00 00 80 16 00 00 00  Normal   Normal
4  1.597760e+09            340    8  FC 03 00 E4 B7 21 FA 3C  Normal   Normal

In [160]:
data_D_attack_2.head()

Timestamp Arbitration_ID  DLC                     Data   Class SubClass
0  1.599046e+09            164    4              00 08 12 5C  Normal   Normal
1  1.599046e+09            340    8  00 00 00 24 D6 01 B5 30  Normal   Normal
2  1.599046e+09            386    8  00 00 00 80 00 40 00 80  Normal   Normal
3  1.599046e+09            485    4              02 00 00 00  Normal   Normal
4  1.599046e+09            130    8  04 80 58 80 00 00 0E 56  Normal   Normal

In [166]:
print('## Shape of Driving state ##')
print('number of data : ', data_D_attack_1.shape[0])
print('number of feature : ', data_D_attack_1.shape[1],'\n')
print('## Data types of Driving state ##')
print(data_D_attack_1.dtypes, '\n')
print('## the number of total IDs in Driving state ##')
print('number of IDs : ',len(data_D_attack_1['Arbitration_ID'].unique()),'\n')

## Shape of Driving state ##
number of data :  806390
number of feature :  6 

## Data types of Driving state ##
Timestamp         float64
Arbitration_ID     object
DLC                 int64
Data               object
Class              object
SubClass           object
dtype: object 

## the number of total IDs in Driving state ##
number of IDs :  81 



In [168]:
print('## Shape of Driving state ##')
print('number of data : ', data_D_attack_2.shape[0])
print('number of feature : ', data_D_attack_2.shape[1],'\n')
print('## Data types of Driving state ##')
print(data_D_attack_2.dtypes, '\n')
print('## the number of total IDs in Driving state ##')
print('number of IDs : ',len(data_D_attack_2['Arbitration_ID'].unique()),'\n')

## Shape of Driving state ##
number of data :  889395
number of feature :  6 

## Data types of Driving state ##
Timestamp         float64
Arbitration_ID     object
DLC                 int64
Data               object
Class              object
SubClass           object
dtype: object 

## the number of total IDs in Driving state ##
number of IDs :  81 



In [171]:
print('## value count of IDs in Driving state attack 1 ##')
data_D_attack_1['Arbitration_ID'].value_counts()

## value count of IDs in Driving state attack 1 ##


000    38521
366    31638
251    31444
220    31412
2B0    31408
       ...  
7D4      270
7DC      266
7C4      265
7D0      255
7D8      246
Name: Arbitration_ID, Length: 81, dtype: int64

In [172]:
print('## value count of IDs in Driving state attack 2 ##')
data_D_attack_2['Arbitration_ID'].value_counts()

## value count of IDs in Driving state attack 2 ##


000    38852
366    35945
220    34789
153    34784
260    34741
       ...  
4A4      277
7C4      265
7D0      260
7DC      259
7CC      236
Name: Arbitration_ID, Length: 81, dtype: int64

 added fake data in attack_D_1 attack_D_2: ['000' '4A4' '7CC' '7D4' '7DC' '7C4' '7D0' '7D8']

In [206]:
print('## the number of attack in Driving state attack 1 ##')
print('Flooding: ',len(data_D_attack_1[data_D_attack_1['SubClass'].isin(['Flooding'])]))
print('Spoofing: ',len(data_D_attack_1[data_D_attack_1['SubClass'].isin(['Spoofing'])]))
print('Replay  : ',len(data_D_attack_1[data_D_attack_1['SubClass'].isin(['Replay'])]))
print('Fuzzing : ',len(data_D_attack_1[data_D_attack_1['SubClass'].isin(['Fuzzing'])]))


## the number of attack in Driving state attack 1 ##
Flooding:  38521
Spoofing:  988
Replay  :  10509
Fuzzing :  22620


In [207]:
print('## the number of attack in Driving state attack 2 ##')
print('Flooding: ',len(data_D_attack_2[data_D_attack_2['SubClass'].isin(['Flooding'])]))
print('Spoofing: ',len(data_D_attack_2[data_D_attack_2['SubClass'].isin(['Spoofing'])]))
print('Replay  : ',len(data_D_attack_2[data_D_attack_2['SubClass'].isin(['Replay'])]))
print('Fuzzing : ',len(data_D_attack_2[data_D_attack_2['SubClass'].isin(['Fuzzing'])]))

## the number of attack in Driving state attack 2 ##
Flooding:  38852
Spoofing:  2891
Replay  :  13266
Fuzzing :  22854


In [211]:
Spoofing_IDs1 = data_D_attack_1[data_D_attack_1['SubClass'].isin(['Spoofing'])]
Spoofing_IDs1['Arbitration_ID'].value_counts()

553    494
366    494
Name: Arbitration_ID, dtype: int64

In [212]:
Spoofing_IDs2 = data_D_attack_2[data_D_attack_2['SubClass'].isin(['Spoofing'])]
Spoofing_IDs2['Arbitration_ID'].value_counts()

553    1448
366    1443
Name: Arbitration_ID, dtype: int64

In [214]:
Spoofing_IDs1[Spoofing_IDs1['Arbitration_ID'].isin(['553'])]

Timestamp Arbitration_ID  DLC                     Data   Class  \
180846  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   
180994  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   
181136  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   
181283  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   
181439  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   
...              ...            ...  ...                      ...     ...   
590546  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   
590691  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   
590836  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   
590984  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   
591135  1.597760e+09            553    8  00 00 00 02 01 00 80 00  Attack   

        SubClass  
180846  Spoofing  
180994  Spoofing  
181136  Spoofing  
181283  Spoofing  
181439  Spoofing  
...          ...  
590546  Spoofing  
590691  Spoofing  
590836  Spoofing  
590984  Spoofing  
591135  Spoofing  

[494 rows x 6 columns]

In [215]:
data_D_attack_1[data_D_attack_1['Arbitration_ID'].isin(['553'])]

Timestamp Arbitration_ID  DLC                     Data   Class  \
151     1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   
633     1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   
1111    1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   
1590    1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   
2072    1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   
...              ...            ...  ...                      ...     ...   
804013  1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   
804492  1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   
804974  1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   
805452  1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   
805931  1.597760e+09            553    8  00 01 19 02 01 00 80 00  Normal   

       SubClass  
151      Normal  
633      Normal  
1111     Normal  
1590     Normal  
2072     Normal  
...         ...  
804013   Normal  
804492   Normal  
804974   Normal  
805452   Normal  
805931   Normal  

[2281 rows x 6 columns]

ID 553, ID 366 are main Flooding target to attack IDS



In [199]:
Replay_IDs1 = data_D_attack_1[data_D_attack_1['SubClass'].isin(['Replay'])]
Replay_IDs1['Arbitration_ID'].value_counts()

130    438
260    438
164    438
251    438
2B0    438
      ... 
043      5
5BE      4
07F      4
5B0      4
57F      2
Name: Arbitration_ID, Length: 73, dtype: int64

In [200]:
Replay_IDs2 = data_D_attack_2[data_D_attack_2['SubClass'].isin(['Replay'])]
Replay_IDs2['Arbitration_ID'].value_counts()

367    560
130    552
329    552
260    552
140    552
      ... 
043      6
5BE      5
07F      5
5B0      5
57F      2
Name: Arbitration_ID, Length: 73, dtype: int64

Replay attack all ID in CAN network ( 73 IDs)

In [201]:
Fuzzing_IDs1 = data_D_attack_1[data_D_attack_1['SubClass'].isin(['Fuzzing'])]
Fuzzing_IDs1['Arbitration_ID'].value_counts()

251    541
140    522
2B0    518
220    507
130    507
      ... 
7D8    235
544    233
44E    233
495    219
500    218
Name: Arbitration_ID, Length: 80, dtype: int64

In [203]:
Fuzzing_IDs2 = data_D_attack_2[data_D_attack_2['SubClass'].isin(['Fuzzing'])]
Fuzzing_IDs2['Arbitration_ID'].value_counts()

220    547
153    534
043    523
251    512
130    505
      ... 
5BE    230
38D    227
495    227
7CC    223
4CB    216
Name: Arbitration_ID, Length: 80, dtype: int64

Fuzzing attacks are all Id + 7 random IDs